internal/util/observability/funccall.go

```diff
@@ -31,17 +31,19 @@ type funcCall struct {
 // FuncCall adds observability to a function call.
 //
 // It should be called at the very beginning of the function,
-// and returned function should be called at exit.
+// and returned [context.CancelFunc] function should be called at exit.
 // The returned function must not be passed or stored.
 // The only valid way to use FuncCall is:
 //
 //	func foo(ctx context.Context) {
-//	    defer FuncCall(ctx)()
-//	    // ...
+//		ctx, cancel := FuncCall(ctx)
+//		defer cancel()
+//		// ...
+//	}
 //
 // For the Go execution tracer, FuncCall creates a new region for the function call
 // and attaches it to the task in the context (or background task).
-func FuncCall(ctx context.Context) func() {
+func FuncCall(ctx context.Context) (context.Context, context.CancelFunc) {
```

Comment:
It would be great to add a test for this function - to call it from the test and then check that everything was stopped correctly, but I don't see a simple way to do it.